README.md
# Skipchunk

[![Pypi](https://img.shields.io/pypi/v/skipchunk.svg)](https://pypi.python.org/pypi/skipchunk)

[![Travis build status](https://img.shields.io/travis/binarymax/skipchunk.svg)](https://travis-ci.org/binarymax/skipchunk)

[![Documentation Status](https://readthedocs.org/projects/skipchunk/badge/?version=latest)](https://skipchunk.readthedocs.io/en/latest/?badge=latest)

Easy search autosuggest with NLP magic.

Out of the box it provides a hassle-free autosuggest for any corpus from scratch, and latent knowledge graph extraction and exploration.

* Free software: MIT License
* Documentation: https://skipchunk.readthedocs.io.

## Install

```bash
pip install skipchunk
python -m spacy download 'en_core_web_lg'
python -m nltk.downloader wordnet
```

You also need to have Solr or Elasticsearch installed and running somewhere!

The current Solr supported version is 8.4.1, but it might work on other versions.

The current Elasticsearch supported version is 7.6.2, but it might work on other versions.

## Use It!

See the ```./example/``` folder for an end-to-end OSC blog load:

### Solr

Start Solr first!  Doesn't work with Solr cloud yet, but we're working on it.
You'll need to start solr using skipchunk's solr_home directory for now.

Then run this: ```python solr-blog-example.py```

### Elasticsearch

Start Elasticsearch first!

Then run this: ```python elasticsearch-blog-example.py```

## Features

* Identifies and groups the noun phrases and verb phrases in a corpus
* Indexes these phrases in Solr or Elasticsearch for a really good out-of-the-box autosuggest
* Structures the phrases as a graph so that concept-relationship-concept can be easily found
* Meant to handle batched updates as part of a full stack search platform

## Library API

### Engine configuration

You need an engine_config, as a dict, to create skipchunk.
The dict must contain the following entries

- host (the fully qualified URL of the engine web API endpoint)
- name (the name of the graph)
- path (the on-disk location of stateful data that will be kept)
- engine_name (either "solr" or "elasticsearch")

#### Solr engine config example

```python
    engine_config_solr = {
        "host":"http://localhost:8983/solr/",
        "name":"osc-blog",
        "path":"./skipchunk_data",
        "engine_name":"solr"
    }
```

#### Elasticsearch engine config example
```python
    engine_config_elasticsearch = {
        "host":"http://localhost:9200/",
        "name":"osc-blog",
        "path":"./skipchunk_data",
        "engine_name":"elasticsearch"
    }
```

### Skipchunk Initialization

When initializing Skipchunk, you will need to provide the constructor with the following parameters

- engine_config (the dict containing search engine connection details)
- spacy_model="en_core_web_lg" (the spacy model to use to parse text)
- minconceptlength=1 (the minimum number of words that can appear in a noun phrase)
- maxconceptlength=3 (the maximum number of words that can appear in a noun phrase)
- minpredicatelength=1 (the minimum number of words that can appear in a verb phrase)
- maxpredicatelength=3 (the maximum number of words that can appear in a verb phrase)
- minlabels=1 (the number of times a concept/predicate must appear before it is recognized and kept.  The lower this number, the more concepts will be kept - so be careful with large content sets!)
- cache_documents=False
- cache_pickle=False

### Skipchunk Methods

- ```tuplize(filename=source,fields=['title','content',...])``` (Produces a list of (text,document) tuples ready for processing by the enrichment.)
- ```enrich(tuples)``` (Enriching can take a long time if you provide lots of text.  Consider batching at 10k docs at a time.)
- ```save``` (Saves to pickle)
- ```load``` (Loads from pickle)

### Graph API

After enrichment, you can then index the graph into the engine

- ```index(skipchunk:Skipchunk)``` (Updates the knowledge graph in the search engine)
- ```delete``` (Deletes a knowledge graph - be careful!)

After indexing, you can call these methods to get autocompleted concepts or walk the knowledge graph

- ```conceptVerbConcepts(concept:str,verb:str,mincount=1,limit=100) -> list``` ( Accepts a verb to find the concepts appearing in the same context)
- ```conceptsNearVerb(verb:str,mincount=1,limit=100) -> list``` ( Accepts a verb to find the concepts appearing in the same context)
- ```verbsNearConcept(concept:str,mincount=1,limit=100) -> list``` ( Accepts a concept to find the verbs appearing in the same context)
- ```suggestConcepts(prefix:str,build=False) -> list``` ( Suggests a list of concepts given a prefix)
- ```suggestPredicates(prefix:str,build=False) -> list``` ( Suggests a list of predicates given a prefix)
- ```summarize(mincount=1,limit=100) -> list``` ( Summarizes a core)
- ```graph(subject:str,objects=5,branches=10) -> list``` ( Gets the subject-predicate-object neighborhood graph for a subject)

## Credits

Developed by Max Irwin, OpenSource Connections https://opensourceconnections.com

All the blog posts contained in the example directory are copyright OpenSource Connections, and may not be used nor redistributed without permission

HISTORY.rst
History
=======

0.1.0 (2019-06-18)
------------------

* Cookie-cutted.

0.9.0 (2020-09-25)
------------------

* First release on PyPI.

1.0.0 (2020-12-10)
------------------

* Stable API.

1.1.0 (2020-12-10)
------------------

* Beta Release.

1.1.1 (2020-12-10)
------------------

* Basic Readme doc.

1.2.2 (2020-12-14)
------------------

* Configset path fix.
* Static assets moved.

2.0.0 (2021-09-09)
------------------

* SpaCy upgraded to 3.1.2.
* Batch multiprocess processing for better throughput

2.0.1 (2021-09-09)
------------------

* Fixed bulk operator

skipchunk/sq.py
import json
import pysolr
import requests
import os
import shutil
import datetime
import jsonpickle
from datetime import date as dt


def pretty(obj):
        print(jsonpickle.encode(obj,indent=2))

## -------------------------------------------
## Java-Friendly datetime string format

def timestamp():
    return datetime.datetime.now().isoformat() + 'Z'

## -------------------------------------------
## Core Admin!
## TODO: These will eventually be abstracted for both Solr and Elasticsearch 

def coreExists(host,name):
    uri = host + 'admin/cores?action=STATUS&core=' + name
    r = requests.get(uri)
    if r.status_code == 200:
        data = r.json()
        if name in data['status'].keys():
            if "name" in data['status'][name].keys():
                if data['status'][name]["name"]==name:
                    return True
    return False


def coreCreate(host,name,path,timeout=10000):

    #Set this to true only when core is created
    success = False

    if not coreExists(host,name):
        try:            
            #Create the core in solr
            uri = host + 'admin/cores?action=CREATE&name='+name+'&instanceDir='+path+'/conf&config=solrconfig.xml&dataDir='+path+'/data'
            r = requests.get(uri)
            if r.status_code == 200:
                success = True
                #Say cheese
                print('Core',name,'created!')
            else:
                print('SOLR ERROR! Core',name,'could not be created! Have a nice day.')
                print(json.dumps(r.json(),indent=2))

        except:
            print('NETWORK ERROR! Could not connect to Solr server on',host,' ... Have a nice day.')

    return success

def coreList(host):
    cores = []
    try:            
        #Lookup all the cores:
        uri = host + 'admin/cores?action=STATUS'
        r = requests.get(uri)
        if r.status_code == 200:
            #Say cheese
            cores = list(r.json()['status'].keys())

        else:
            print('SOLR ERROR! Cores could not be listed! Have a nice day.')
            print(json.dumps(r.json(),indent=2))

    except:
        print('NETWORK ERROR! Could not connect to Solr server on',host,' ... Have a nice day.')
    
    return cores

## -------------------------------------------
## Indexing!

def indexableGroups(groups,contenttype='concept'):
    """ Generates concept records for Solr, similar to how ES Bulk indexing
        uses a generator to generate bulk index/update actions """
    createtime = timestamp()
    for group in groups:
        key = group.key
        for label in group.labels:
            snippetid = str(label.docid) + '_' + str(label.sentenceid)
            labelid = group.key + '_' + snippetid
            #print("Indexing %s" % labelid)

            record = {
                'id' : labelid,
                'key' : key,
                'idiom' : label.idiom,
                'label' : label.label,
                'length' : label.length,
                'start' : label.start,
                'end' : label.end,
                'docid' : label.docid,
                'sentenceid' : snippetid,
                'sentencenum_s' : label.sentenceid,
                'objectof' : label.objectOf,
                'subjectof' : label.subjectOf,
                'contenttype' : contenttype,
                'createtime': createtime,
                #Group-level values
                'preflabel' : group.preflabel,
                'prefcount' : group.prefcount,
                'total' : group.total
            }

            #For separate concept/predict behavior (such as autosuggest)
            if contenttype == 'concept':
                record['conceptlabel'] = label.label
            else:
                record['predicatelabel'] = label.label

            ##Generator for the record
            yield record

## -------------------------------------------

def indexableLabels(labels,contenttype='concept'):
    """ Generates labels for indexing, similar to how ES Bulk indexing
        uses a generator to generate bulk index/update actions """

    createtime = timestamp()
    for key in labels.keys():
        for label in labels[key]:
            labelid = key + '_' + str(label.docid) + '_' + str(label.sentenceid)

            #print("Indexing %s" % labelid)

            record = {
                'id' : labelid,
                'key' : key,
                'idiom' : label.idiom,
                'label' : label.label,
                'length' : label.length,
                'start' : label.start,
                'end' : label.end,
                'docid' : label.docid,
                'sentenceid' : label.sentenceid,
                'objectof' : label.objectOf,
                'subjectof' : label.subjectOf,
                'contentType' : contenttype,
                'createtime': creattime
            }

            if contenttype == 'concept':
                record['conceptlabel'] = label.label
            else:
                record['predicatelabel'] = label.label

            ##Generator for the record
            yield record

## -------------------------------------------
## Querying shortcuts

def printAggregate(agg):
    if agg:
        for f in agg:
            print(f["label"],f["count"])

def printSuggest(sugg):
    if sugg:
        for f in sugg:
            print(f["term"],f["weight"])

def cleanTerm(term):
    return term.replace('/','\\/').replace('"','')

## -------------------------------------------
## Facets

def aggregateQuery(field,mincount=1,limit=100):
    return {
        "facet.field":field,
        "facet.limit":limit,
        "facet.mincount":mincount,
        "facet":"on"
    }

def parseAggregate(field,res):
    facets = None
    if (res.facets) and ("facet_fields" in res.facets.keys()) and (field in res.facets["facet_fields"].keys()):
        labels = res.facets["facet_fields"][field]
        facets = []
        for i in range(len(labels)//2):
            facets.append({"label":labels[i*2],"count":labels[i*2+1]})

    return facets

## -------------------------------------------
## Suggestions

def suggest(prefix,dictionary="conceptLabelSuggester",count=25,build=False):
    s = {
        "suggest":"true",
        "suggest.q":prefix,
        "suggest.count":count,
        "suggest.dictionary":dictionary
    }
    if build:
        s["suggest.build"] = "true"

    return s


##==========================================================
# MAIN API ENTRY POINT!  USE THIS!

class SkipchunkQuery():

    ## -------------------------------------------
    # Accepts a skipchunk object to index the required data in Solr
    def index(self,skipchunk,timeout=10000):

        isCore = coreExists(self.host,skipchunk.name)
        if not isCore:
            isCore = coreCreate(self.host,skipchunk.name,skipchunk.solr_data)

        if isCore:
            indexer = pysolr.Solr(self.solr_uri, timeout=timeout)
            indexer.add(list(indexableGroups(skipchunk.predicategroups,contenttype="predicate")),commit=True)
            indexer.add(list(indexableGroups(skipchunk.conceptgroups,contenttype="concept")),commit=True)

    def cores(self):
        return coreList(self.host)

    def changeCore(self,name):
        if coreExists(self.host,name):
            self.name = name
            self.solr_uri = self.host + name
            self.select_handler = pysolr.Solr(self.solr_uri)
            self.suggest_handler = pysolr.Solr(self.solr_uri, search_handler='/suggest')            
            return True
        return False

    ## -------------------------------------------
    # Uses a streaming expression to fill in missing prefLabels after indexing
    def makePrefLabels():    
        # TODO: Write this
        return 0

    ## -------------------------------------------
    # Accepts a verb to find the concepts appearing in the same context

    def conceptVerbConcepts(self,concept,verb,mincount=1,limit=100):
        
        subject = cleanTerm(concept)
        verb = cleanTerm(verb)
        objects = []
        subjects = []

        # Get all the docid and sentenceid pairs that contain both the concept AND verb
        #    http://localhost:8983/solr/osc-blog/select?fl=docid%2Csentenceid&fq=subjectof%3A%22be%22%20OR%20objectof%3A%22be%22&q=label%3A%22open%20source%22&rows=1000
        q="*"
        fl="sentenceid"
        fq=["(subjectof:\""+verb+"\" OR objectof:\""+verb+"\")","preflabel:\""+subject+"\""]
        rows=10000
        res = self.select_handler.search(q=q,fq=fq,fl=fl,rows=rows)

        if len(res.docs)>0:

            # Get all the other concepts that exist in those docid and sentenceid pairs
            #   http://localhost:8983/solr/osc-blog/select?fl=*&fq=-preflabel%3A%22open%20source%22&fq=sentenceid%3A17%20AND%20docid%3Aafee4d71ccb3e19d36ee2cfddd6da618&q=contenttype%3Aconcept&rows=100
            sentences = []
            for doc in res.docs:
                sentences.append("sentenceid:"+doc["sentenceid"])
            sentenceids = " OR ".join(sentences)

            q="*"
            fq=[
                "-preflabel:\""+subject+"\"",
                #"(subjectof:\""+verb+"\" OR objectof:\""+verb+"\")",
                "(" + sentenceids + ")"
            ]
            field = "preflabel"

            facet = aggregateQuery(field,mincount,limit)
            res   = self.select_handler.search(q=q,fq=fq,rows=rows,**facet)
            objects = parseAggregate(field,res)
        
        return objects

    ## -------------------------------------------
    # Accepts a verb to find the concepts appearing in the same context

    def conceptsNearVerb(self,verb,mincount=1,limit=100):
        
        verb = cleanTerm(verb)

        field = "preflabel"
        q = "*:*"
        fq = "objectof:"+verb+" OR subjectof:"+verb

        facet = aggregateQuery(field,mincount,limit)
        res = self.select_handler.search(q=q,fq=fq,rows=0,**facet)
        aggregate = parseAggregate(field,res)

        return aggregate

    ## -------------------------------------------
    # Accepts a concept to find the verbs appearing in the same context

    def verbsNearConcept(self,concept,mincount=1,limit=100):

        concept = cleanTerm(concept)

        field = "subjectof"
        q = concept
        facet = aggregateQuery(field,mincount,limit)
        res = self.select_handler.search(q=q,rows=0,**facet)
        subjectofs = parseAggregate(field,res)

        field = "objectof"
        q = concept
        facet = aggregateQuery(field,mincount,limit)
        res = self.select_handler.search(q=q,rows=0,**facet)
        objectofs = parseAggregate(field,res)


        return subjectofs+objectofs

    ## -------------------------------------------
    # Suggests a list of concepts given a prefix

    def suggestConcepts(self,prefix,build=False):
        dictionary="conceptLabelSuggester"
        query = suggest(prefix,dictionary=dictionary,count=5,build=build)
        res = self.suggest_handler.search(q=prefix,**query)
        return res.raw_response["suggest"][dictionary][prefix]["suggestions"]

    ## -------------------------------------------
    # Suggests a list of predicates given a prefix

    def suggestPredicates(self,prefix,build=False):

        objectofSuggestions = []
        subjectofSuggestions = []

        dictionary1="objectofLabelSuggester"
        query1 = suggest(prefix,dictionary=dictionary1,count=5,build=build)
        res1 = self.suggest_handler.search(q=prefix,**query1)
        if dictionary1 in res1.raw_response["suggest"]:
            objectofSuggestions = res1.raw_response["suggest"][dictionary1][prefix]["suggestions"]

        dictionary2="subjectofLabelSuggester"
        query2 = suggest(prefix,dictionary=dictionary2,count=5,build=build)
        res2 = self.suggest_handler.search(q=prefix,**query2)
        if dictionary2 in res2.raw_response["suggest"]:
            subjectofSuggestions = res2.raw_response["suggest"][dictionary2][prefix]["suggestions"]

        return objectofSuggestions


    ## -------------------------------------------
    # Summarizes a core

    def summarize(self,mincount=1,limit=100):

        q  = "*"
        fq = "contenttype:concept"
        field = "preflabel"

        facet = aggregateQuery(field,mincount,limit)
        res = self.select_handler.search(q=q,fq=fq,rows=0,**facet)
        concepts = parseAggregate(field,res)

        q2  = "*"
        fq2 = "contenttype:predicate"
        field2 = "preflabel"

        facet2 = aggregateQuery(field2,mincount,limit)
        res2 = self.select_handler.search(q=q2,fq=fq2,rows=0,**facet2)
        predicates = parseAggregate(field2,res2)

        return concepts,predicates

    ## -------------------------------------------
    # Gets the subject-predicate-object graph for a subject

    def graph(self,subject,objects=5,branches=10):

        tree = []

        verbs = self.verbsNearConcept(subject)[0:branches]
        branch = {
            "label":subject,
            "labeltype":"subject",
            "relationships":[]
        }
        for verb in verbs:
            v = verb["label"]
            predicate = {
                "label":v,
                "weight":verb["count"],
                "labeltype":"predicate",
                "relationships":[]
            }
            cvc = self.conceptVerbConcepts(subject,v,limit=objects)
            for o in cvc:
                predicate["relationships"].append({
                    "label":o["label"],
                    "weight":o["count"],
                    "labeltype":"object",
                    "relationships":[]
                })
            branch["relationships"].append(predicate)

        tree.append(branch)

        return tree

    ## -------------------------------------------
    # Pretty-prints a graph walk of all suggested concepts and their verbs given a starting term prefix

    def explore(self,term,contenttype="concept",build=False,quiet=False,branches=10):

        tree = []

        if contenttype == "concept":
            if not quiet:
                print("\n=================FINDING '"+term+"' ==================")
            for s in self.suggestConcepts(term,build=build):
                verbs = self.verbsNearConcept(s["term"])[0:branches]
                c = s["term"]
                branch = {
                    "label":c,
                    "weight":s["weight"],
                    "labeltype":"subject",
                    "relationships":[]
                }
                for verb in verbs:
                    v = verb["label"]
                    predicate = {
                        "label":v,
                        "weight":verb["count"],
                        "labeltype":"predicate",
                        "relationships":[]
                    }
                    cvc = self.conceptVerbConcepts(c,v,limit=5)
                    for o in cvc:
                        predicate["relationships"].append({
                            "label":o["label"],
                            "weight":o["count"],
                            "labeltype":"object",
                            "relationships":[]
                        })
                    branch["relationships"].append(predicate)

                tree.append(branch)

                if not quiet:
                    lst = ', '.join([a["label"]+"("+str(a["count"])+")" for a in verbs])
                    print("")
                    print(s["term"] + " (" + str(s["weight"]) + ")")
                    print('\t',lst)
                    print("")
                    print('----------------------------------------------------')
        
        elif contenttype == "predicate":
            if not quiet:
                print("\n=================FINDING '"+term+"' ==================")
            for s in self.suggestPredicates(term,build=build):
                concepts = self.conceptsNearVerb(s["term"])[0:branches]
                tree[s["term"]] = concepts
                if not quiet:
                    lst = ', '.join([a["label"]+"("+str(a["count"])+")" for a in verbs])
                    print("")
                    print(s["term"] + " (" + str(s["weight"]) + ")")
                    print('\t',lst)
                    print("")
                    print('----------------------------------------------------')

        return tree

    ## -------------------------------------------
    # host:: the url of the solr server
    # name:: the name of the solr core
    def __init__(self,host,name):
        self.host = host
        self.name = name
        self.solr_uri = host + name
        self.select_handler = pysolr.Solr(self.solr_uri)
        self.suggest_handler = pysolr.Solr(self.solr_uri, search_handler='/suggest')


##==========================================================
# Command line explorer!

if __name__ == "__main__":
    import sys
    import jsonpickle

    def pretty(obj):
        print(jsonpickle.encode(obj,indent=2))

    if len(sys.argv)<3:
        print('python sq.py <solr_core_name> <concept|predicate> <term>')

    else:
        i=1
        if sys.argv[0]=='python':
            i=2

        if len(sys.argv)<i+3:
            print('python sq.py <solr_core_name> <concept|predicate> <term>')
        else:
            core = sys.argv[i]
            kind = sys.argv[i+1]
            term = sys.argv[i+2]

            if coreExists('http://localhost:8983/solr/',core):
                sq = SkipchunkQuery('http://localhost:8983/solr/',core)
                #sq.explore(term,contenttype=kind,build=False)
                sq.graph(term)
            else:
                print('Core "',core,'" does not exists on localhost')
